### PR TITLE
feat(examples): add mnist example

### LINKED ISSUE #99: perf(tensor/cpu): optimize performance
# Here is Claudes analysis result for benchmark MNIST run.

## How to Profile

```bash
cd examples/04-mnist
go run main.go *.go          # generates cpu.prof
go tool pprof -top cpu.prof  # text summary of hotspots
go tool pprof -http=:8080 cpu.prof  # flame graph in browser
```

---

## Root Cause — Nested `[]any` Data Structure

All bottlenecks trace back to one architectural decision: tensors store data as nested `[]any` instead of a flat `[]float64`.

```go
// tensor/internal/cputensor/initializers.go:8–27
func (t *CPUTensor) initWith(initFunc initializerFunc) {
    fill = func(dims []int, data *any) {
        rows := make([]any, dims[0])   // heap alloc per row per op
        for i := range rows { fill(dims, &rows[i]) }
        *data = rows
    }
}
```

Consequences:
- **Every element access requires a type assertion**: `data.([]any)[i].(float64)`
- **Cache thrashing**: each row is a separate heap pointer; traversing a 100×784 matrix chases 100 separate pointers
- **Heap fragmentation**: O(rows × ndim) allocations per tensor operation

---

### #1 — `matMulDataOf2DInputs` (`tensor/internal/cputensor/operators.go:297`) — ~50–70% CPU

The inner loop of the naive 3-loop matmul does 4 type assertions per multiply:

```go
for p := range n {
    rim1 := m1[i].([]any)      // type assertion + potential cache miss
    rpm2 := m2[p].([]any)      // type assertion
    eipm1 := rim1[p].(float64) // type assertion
    epjm2 := rpm2[j].(float64) // type assertion
    eij += eipm1 * epjm2
}
```

Per training batch:
- FC1 forward: 100 × 784 × 500 = **39.2M inner iterations**
- FC1 backward (weight grad): 100 × 500 × 784 = **39.2M** more
- FC2 forward + backward: additional ~5M

No cache blocking → random-access pattern for matrix B → L1/L2 thrashing.

### #2 — `applyUnaryFuncOnTensorElemWise` / `applyBinaryFuncOnTensorsElemWise` (`operators.go:173–228`) — ~15–25% CPU

Every `add`, `sub`, `mul`, `scale`, `exp`, `log`, `relu` call allocates a new `[]any` per row:

```go
rRows := make([]any, dims[0])   // heap alloc on every call, every row
```

Called ~20+ times per training step for Adam alone (8+ ops × 2 layers of parameters).

### #3 — `reduceDimUsingTensorFunc` + `linearElemGeneratorWithReducedDim` (`reducers.go:129`) — ~5–10% CPU

`FC.forward` calls `y.SumAlong(2)` after matmul. This creates a new `CPUTensor` via `t.slice(state)` for every output element — allocating 500 tensors per batch forward pass through FC1.

### #4 — `transpose` + `reshape` (`shape_modifiers.go`) — ~5–10% CPU

FC backward triggers `Transpose()` on both weight and input. Transpose materializes a full copy through `initWith` + element generator. No view semantics — every shape op copies all data.

### #5 — `accumulateGrad` allocations (`tensor/internal/gradtrack/back_propagation.go:57`)

Every non-first gradient accumulation calls `gctx.gradient.Add(grad)` → new tensor allocation. Should be in-place `+=`.

---

#### O1. Replace `[]any` with flat `[]float64` storage
**Impact: 5–20× overall speedup**

The single highest-leverage change.

```go
// current
type CPUTensor struct { data any; dims []int }

// proposed
type CPUTensor struct { data []float64; dims []int; strides []int }
```

Benefits:
- All element access: `data[i*stride0 + j*stride1]` — no type assertions, contiguous memory
- 100×784 matrix = one contiguous 62KB block (fits in L2 cache, no pointer chasing)
- Enables view semantics: reshape/transpose change `dims`/`strides`, share the same `data` slice

#### O2. Use BLAS for matrix multiply (`gonum/blas64`)
**Impact: 10–100× matmul speedup (uses AVX-512 + multi-threading)**

`gonum.org/v1/gonum` is already declared in `go.mod`. With flat storage:

```go
import "gonum.org/v1/gonum/blas/blas64"

blas64.Gemm(blas.NoTrans, blas.Trans, 1.0, a, b, 0.0, c)
```

Replaces the pure-Go triple loop in `matMulDataOf2DInputs`. Requires O1 first.

---

#### O3. In-place gradient accumulation
**Impact: eliminates 1 tensor alloc per gradient per backward step**

```go
// current (back_propagation.go:61)
acc, err := gctx.gradient.Add(grad)
gctx.gradient = acc

// proposed
gctx.gradient.AddInPlace(grad)  // loop over []float64 directly
```

Independent of the storage refactor — can be done now.

#### O4. View semantics for reshape/transpose
**Impact: eliminates all copy allocations for shape ops**

With flat storage + strides, `reshape` returns a tensor sharing the same `data` slice with new `dims/strides`. `transpose` swaps `strides[n-2]` and `strides[n-1]`. Zero data copied.

#### O5. Fused FC forward (replace UnSqueeze + MatMul + SumAlong + Add)
**Impact: eliminates 3 of 4 intermediate tensor allocations per FC forward**

Current `FC.forward` (`component/layers/fc.go:76`):
```
UnSqueeze(W) → MatMul → SumAlong(2) → Add(b)   // 4 ops, 4+ allocations
```

With BLAS this becomes a single `dgemm` call (`Y = X·Wᵀ + b`), eliminating the unsqueeze, the sum-along, and the bias add as separate steps. Requires O1 + O2.

#### O6. Cache-blocked naive matmul (without BLAS dependency)
**Impact: 2–4× matmul speedup by improving L1 cache reuse**

Even without switching to flat storage, blocking the B-matrix access pattern helps:

```go
const blockSize = 32
for ii := 0; ii < m; ii += blockSize {
    for kk := 0; kk < n; kk += blockSize {
        for jj := 0; jj < k; jj += blockSize {
            // inner 3-loop capped at block boundaries
        }
    }
}
```

---

#### O7. Fused numerically-stable softmax (single pass)
**Impact: 4 allocations → 1; also fixes potential overflow**

Current softmax calls `Exp()` → `SumAlong()` → broadcast → `Div()` — 4 separate passes and allocations. A fused kernel:
1. Find `max(x)` for numerical stability (1 pass)
2. Compute `exp(x − max)` and accumulate `sum` simultaneously (1 pass)
3. Divide by `sum` in-place (1 pass)

All three steps write into a single pre-allocated output buffer.

#### O8. Preallocate gradient tensors
**Impact: eliminates ~N_params allocations per backward pass**

Allocate zero-filled gradient tensors at model construction time. During backprop, zero them (via memset-equivalent on `[]float64`) and reuse.

#### O9. Adam optimizer in-place updates
**Impact: eliminates 8+ tensor allocations per parameter per step**

Adam currently chains 8+ tensor ops (`scale` → `add` → `pow` → `div` → `sub`), each creating a new tensor. A single fused loop over `[]float64` slices does all updates in one pass.

---

#### O10. Parallel batch processing
**Impact: near-linear scaling up to num_cores**

Forward pass samples within a batch are independent. With flat storage and goroutines, process N samples in parallel using `sync.WaitGroup`. Most effective for the FC forward pass.

#### O11. Parallel element-wise ops
**Impact: near-linear scaling for large tensors**

For tensors over a threshold (~10K elements), split the flat `[]float64` into chunks and process with goroutines. Below the threshold, single-threaded is faster (avoids goroutine overhead).

---

## Summary Table

| ID  | Change                              | Effort | Expected Speedup        | Prerequisite |
|-----|-------------------------------------|--------|-------------------------|--------------|
| O1  | Flat `[]float64` storage            | High   | 5–20× overall           | —            |
| O2  | BLAS matmul (`gonum/blas64`)        | Medium | 10–100× matmul          | O1           |
| O3  | In-place gradient accumulation      | Low    | 10–20% backprop         | —            |
| O4  | View semantics reshape/transpose    | Medium | 5–15% shape ops         | O1           |
| O5  | Fused FC forward (BLAS dgemm)       | Medium | 10–30% FC layer         | O1, O2       |
| O6  | Cache-blocked naive matmul          | Medium | 2–4× matmul             | —            |
| O7  | Fused softmax kernel                | Low    | 5–10%                   | —            |
| O8  | Preallocate gradient tensors        | Low    | 5–10% backward          | —            |
| O9  | Adam in-place updates               | Medium | 5–15% optimizer step    | —            |
| O10 | Parallel batch forward              | Medium | ~N_cores×               | O1           |
| O11 | Parallel element-wise ops           | Low    | ~N_cores× large tensors | —            |

**Recommended order**: Start with O3 and O7 (low effort, independent). Then tackle O1 as the foundational refactor that unlocks O2, O4, O5, O10.

---

## Verification

```bash
# Baseline measurement
time go run main.go *.go
go tool pprof -top cpu.prof

# After each optimization
time go run main.go *.go         # compare wall-clock time
go tool pprof -top cpu.prof      # verify the hotspot dropped
# Confirm accuracy hasn't regressed: should print Accuracy: ~0.97
```

### PR DESCRIPTION
**Pull request by sahandsafizadeh**
## Feat(examples): add mnist example

The MNIST example drove major improvements across the whole project.

- **Before:** training produced 0% accuracy (`NaN` loss), or ~10% at best, with each epoch taking ~22 minutes.
- **After:** final accuracy reaches **98%** at **~35s per epoch**.

## Fix(tensor/gradtrack)!: apply Kahn's method

- Use a BFS-style Kahn algorithm: a node's children are enqueued for the backward pass only after that node's gradient is fully finalized. This requires a single forward pass to count each node's consumers.
- Some networks run backprop over a single node more than once, so gradients from separate runs are accumulated on top of each other via snapshots.
- Skipping `bpDirty` significantly worsened the memory leak and caused `SIGKILL` mid-training. An `ensureGradSafeToAccumulate` guard now runs at the start of every gradient accumulation and panics to prevent this.
- This fix adds several new backpropagation test scenarios.

## Fix(tensor/gradtrack): broadcasting gradient

- Previously we incorrectly **averaged** when computing the gradient of a broadcast operation.
- Broadcasting is equivalent to copying a tensor, so its gradient is the **sum** of the next layer's gradients.
- This fix optimizes the MNIST loss and reaches **96% accuracy in the first epoch**.

## Fix(component/layers): FC implementation — weight tensor of shape `[in, out]`

- Matches the standard linear-layer formulation from the literature.
- The weight parameter is now a matrix instead of a vector.
- The additional trainable parameters bring model accuracy in line with real experiments.
- The new implementation is also more computationally efficient: it removes the oversized intermediate tensor from the `matmul`.
- The input count is stored privately for simpler tracking and weight validation.

## Fix(component/activations): make Softmax numerically stable

- The previous naive implementation overflowed quickly. Subtracting the per-dimension maximum now prevents infinite values.

## Fix(tensor/cputensor): make reducer operations incremental and numerically stable

- A new reducer interface unifies reducing all dimensions and reducing a specific dimension.
- More importantly, this design enables numerically stable reductions such as mean and variance.

## Perf(tensor/cputensor): redesign the CPU tensor

- The nested `any` / `[]any` structure was the main bottleneck behind many performance issues. Storage is migrated to a flat `[]float64` plus dimensions and strides.
- This yields far fewer memory allocations and less cache thrashing.
- Storing strides means most reshape operations (e.g. `transpose`) allocate no new tensors and instead reuse the immutable underlying data of the operand.
- Most functions use the `at` method and follow the `newTensorWithElementWiseInit` index-update pattern for maintainability, at the cost of a constant `len(dims)` factor.
- **Follow-ups:** `matmul` still needs investigation (it does not yet use the `at` / index pattern), as does `broadcast` (possibly via virtual views that avoid allocating new tensors).

## Perf: improve all previous examples in both accuracy and speed

- Following the FC, Softmax, backprop, and broadcast-gradient fixes, proven models are used for the examples (Boston Housing, Iris, Breast Cancer), and results now match the accuracies reported by each model's source.
- All model training/evaluation times are now real-time.

## Perf(model): accumulate the extracted float64 loss instead of adding tensors

- The loss function is tracked, so accumulating loss tensors across epochs builds ever-larger gradient-tracked graphs.
- The loss value is now extracted once, at the end of `trainStep`.
- As a result, the epoch logger no longer depends on tensors and only has to write a float64 value.

## Devops: ignore profiling output files

- The performance work in this PR relied heavily on Go's profiling tools, so pprof outputs are now git-ignored.

## Refactor: introduce `qeep/internal` and move `model/internal/queue` there

- The queue is required by backprop (tensor path) and forward prop (model framework) in different places, so this high-performance implementation is now shared across the project.

### CODE REVIEW

**Review by sahandsafizadeh — COMMENTED** [review]
approved
